bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/TraversePageHandler.java

```diff
@@ -28,30 +28,84 @@
  * @since 3.5
  */
 public class TraversePageHandler extends WidgetMethodHandler {
-
 	/**
 	 * The parameters for traverse(int).
 	 */
 	private static final Class<?>[] METHOD_PARAMETERS = { int.class };
-
 	@Override
 	public final Object execute(final ExecutionEvent event) {
 		Control focusControl = Display.getCurrent().getFocusControl();
 		if (focusControl != null) {
-			int traversal = "next".equals(methodName) ? SWT.TRAVERSE_PAGE_NEXT : SWT.TRAVERSE_PAGE_PREVIOUS; //$NON-NLS-1$
+			boolean forward = "next".equals(methodName); //$NON-NLS-1$
+			int traversal = getTraversalDirection(forward);
 			Control control = focusControl;
+
 			do {
+				if (control instanceof CTabFolder folder && isFinalItemInCTabFolder(folder, forward)
+						&& !areHiddenItems(folder)) {
+					loopToSecondToFirstItemInCTabFolder(folder, forward);
```

Comment:
"loopToSecondToFirstItemInCTabFolder" is not a desirable name. Are you referring to the "second to last" item or to the "second" item?

Reply:
done